diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/SmartCastRule.kt

```diff
@@ -0,0 +1,286 @@
+                    val afterDotPart = dotExpr.text.split(".")[1]
+                    val text = "${it.identifier}.$afterDotPart"
+                    dotExpr.treeParent.addChild(KotlinParser().createNode(text), dotExpr)
+                    dotExpr.treeParent.removeChild(dotExpr)
+                }
+            }
+        }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun handleWhenCondition(node: ASTNode) {
+        /*
+            Check if there is WHEN_CONDITION_IS_PATTERN. If so delete 'as' in it's block
+            or call expression if it doesn't have block
+         */
+
+        val identifier = node.getFirstChildWithType(REFERENCE_EXPRESSION)?.text
+
+        node.getAllChildrenWithType(WHEN_ENTRY).forEach {
+            if (it.hasChildOfType(WHEN_CONDITION_IS_PATTERN) && identifier != null) {
+                val type = it.getFirstChildWithType(WHEN_CONDITION_IS_PATTERN)!!
+                        .getFirstChildWithType(TYPE_REFERENCE)!!.text
+
+                val callExpr = it.getFirstChildWithType(CALL_EXPRESSION)!!
+                val blocks = listOf(IsExpressions(identifier, type))
+
+                handleThenBlock(callExpr, blocks)
+            }
+        }
+    }
+
+    private fun KtNameReferenceExpression.hasLocalDeclaration(): KtProperty? = parents
+            .mapNotNull { it as? KtBlockExpression }
+            .first()
+            .let { blockExpression ->
+                blockExpression
+                        .statements
+                        .takeWhile { !it.isAncestor(this, true) }
+                        .mapNotNull { it as? KtProperty }
+                        .find {
+                            it.isLocal &&
+                                    it.hasInitializer() &&
+                                    it.name?.equals(getReferencedName())
+                                    ?: false
+                        }
+            }
+
+    private fun collectLocalPropertiesWithUsages(node: ASTNode) = node
+            .findAllVariablesWithUsages { propertyNode ->
+                propertyNode.name != null
+            }
+
+    /**
+     * Gets references, which contains is or as keywords
+     * @returns Map of property and list of expressions
+     */
+    private fun collectReferenceList(propertiesToUsages: Map<KtProperty, List<KtNameReferenceExpression>>)
+            : Map<KtProperty, List<KtNameReferenceExpression>> {
+        val properMap: MutableMap<KtProperty, List<KtNameReferenceExpression>> = mutableMapOf()
+        propertiesToUsages.keys.forEach {
+            val list: MutableList<KtNameReferenceExpression> = mutableListOf()
+            propertiesToUsages.getValue(it).forEach { entry ->
+                if (entry.parent.node.elementType == IS_EXPRESSION
+                        || entry.parent.node.elementType == BINARY_WITH_TYPE) {
+                    list.add(entry)
+                }
+            }
+            properMap[it] = list
+        }
```

Comment:
I think `propertiesToUsages.mapValues{ (_, value) -> value.filter { ... } }` will be shorter and more readable.